backend/oracleobjectstorage/multipart.go

```diff
@@ -104,231 +86,351 @@ func (o *Object) uploadParts(
+	req := objectstorage.CommitMultipartUploadRequest{
+		NamespaceName: common.String(w.f.opt.Namespace),
+		BucketName:    w.bucket,
+		ObjectName:    w.key,
+		UploadId:      w.uploadId,
 	}
-
 	// sort the completed parts by part number
-	sort.Slice(parts, func(i, j int) bool {
-		return *parts[i].PartNum < *parts[j].PartNum
+	sort.Slice(w.partsToCommit, func(i, j int) bool {
+		return *w.partsToCommit[i].PartNum < *w.partsToCommit[j].PartNum
 	})
-
+	req.PartsToCommit = w.partsToCommit
 	var resp objectstorage.CommitMultipartUploadResponse
-	resp, err = o.commitMultiPart(ctx, uploadID, parts)
+	err = w.f.pacer.Call(func() (bool, error) {
+		resp, err = w.f.srv.CommitMultipartUpload(ctx, req)
+		// if multipart is corrupted, we will abort the uploadId
+		if isMultiPartUploadCorrupted(err) {
+			fs.Debugf(w.o, "multipart uploadId %v is corrupted, aborting...", *w.uploadId)
+			_ = w.Abort(ctx)
+			return false, err
+		}
+		return shouldRetry(ctx, resp.HTTPResponse(), err)
+	})
 	if err != nil {
 		return err
 	}
-	fs.Debugf(o, "multipart upload %v committed.", uploadID)
-	hashOfHashes := md5.Sum(md5s)
-	wantMultipartMd5 := base64.StdEncoding.EncodeToString(hashOfHashes[:]) + "-" + strconv.Itoa(len(parts))
+	w.eTag = *resp.ETag
+	hashOfHashes := md5.Sum(w.md5s)
+	wantMultipartMd5 := fmt.Sprintf("%s-%d", base64.StdEncoding.EncodeToString(hashOfHashes[:]), len(w.partsToCommit))
 	gotMultipartMd5 := *resp.OpcMultipartMd5
 	if wantMultipartMd5 != gotMultipartMd5 {
-		fs.Errorf(o, "multipart upload corrupted: multipart md5 differ: expecting %s but got %s", wantMultipartMd5, gotMultipartMd5)
+		fs.Errorf(w.o, "multipart upload corrupted: multipart md5 differ: expecting %s but got %s", wantMultipartMd5, gotMultipartMd5)
 		return fmt.Errorf("multipart upload corrupted: md5 differ: expecting %s but got %s", wantMultipartMd5, gotMultipartMd5)
 	}
-	fs.Debugf(o, "multipart upload %v md5 matched: expecting %s and got %s", uploadID, wantMultipartMd5, gotMultipartMd5)
+	fs.Debugf(w.o, "multipart upload %v md5 matched: expecting %s and got %s", *w.uploadId, wantMultipartMd5, gotMultipartMd5)
 	return nil
 }
 
-// commits the multipart upload
-func (o *Object) commitMultiPart(ctx context.Context, uploadID string, parts []*objectstorage.CommitMultipartUploadPartDetails) (resp objectstorage.CommitMultipartUploadResponse, err error) {
-	bucketName, bucketPath := o.split()
-	req := objectstorage.CommitMultipartUploadRequest{
+func isMultiPartUploadCorrupted(err error) bool {
+	if err == nil {
+		return false
+	}
+	// Check if this oci-err object, and if it is multipart commit error
+	if ociError, ok := err.(common.ServiceError); ok {
+		// If it is a timeout then we want to retry that
+		if ociError.GetCode() == "InvalidUploadPart" {
+			return true
+		}
+	}
+	return false
+}
+
+func (w *objectChunkWriter) Abort(_ context.Context) error {
+	fs.Debugf(w.o, "Cancelling multipart upload")
+	err := w.o.fs.abortMultiPartUpload(
+		context.Background(),
```

Comment:
You should be passing the `ctx` passed into `Abort(ctx)` here I think?

Reply:
yes i thought so too, but if passed in ctx is cancelled, the abort should proceed or not, I guess it should stop as well, so passing in the ctx as you suggested.